core/controllers/load_functions.py

```diff
@@ -0,0 +1,257 @@
+import pandas as pd
+from sqlalchemy import desc, func
+
+from core.const import SUBMISSION_ID_FORMAT
+from core.controllers.mappings import DataMapping
+from core.db import db
+from core.db.entities import Organisation, Programme, Project, Submission
+
+
+def load_programme_ref(workbook: dict[str, pd.DataFrame], mapping: DataMapping):
+    """Loads data into the 'Programme_Ref' table.
+
+    If the programme exists in a previous round, the incoming programme's information supersedes the existing programme.
+    Data is added to the session via this method, but not committed.
+
+    :param workbook: a dictionary of DataFrames of table data to be inserted into the db.
+    :param mapping: the mapping of the relevant DataFrame to its attributes as they appear in the db.
+    """
+    reporting_round = int(workbook["Submission_Ref"]["Reporting Round"].iloc[0])
+    programme_id = workbook["Programme_Ref"]["Programme ID"].iloc[0]
+    # if added before or during this round, get the programme entity to merge (update)
+    programme_exists_previous_round = (
```

Comment:
maybe move this db query stuff into a separate function, then it's easier to test/mock, and scope definition / function roles are clearer?

Reply:
It could even go in core/db/queries, just as a simple function that only has the query in it. That way, all DB query code is in the same place, and is also easier to re-use etc.

Reply:
Can do